pkg/jsonpath/expand_test.go

```diff
@@ -0,0 +1,388 @@
+/*
+Copyright 2020 The Tekton Authors
+	"github.com/tektoncd/pipeline/pkg/substitution"
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package jsonpath
+
+import (
+	"fmt"
+	"reflect"
+	"testing"
+)
+
+func hasVariables(text string) bool {
+	matches := expandRE.FindAllString(text, -1)
+	for _, m := range matches {
+		if m != "$$" {
+			return true
+		}
+	}
+	return false
+}
+func TestHasVariables(t *testing.T) {
+	tests := []struct {
+		text     string
+		expected bool
+	}{
+		{"", false},
+		{"abc", false},
+		{"$(foo.bar)", true},
+		{"abc $(foo.bar)", true},
+		{"$$(foo.bar)", false},
+		{"$$(foo.bar) $(foo.bar)", true},
+		{"$$$(foo.bar)", true},
+	}
+
+	for _, test := range tests {
+		t.Run(test.text, func(t *testing.T) {
+			if hasVariables(test.text) != test.expected {
+				t.Errorf("hasVariables() for  %v unexpectedly return %v.", test.text, !test.expected)
+			}
+		})
+	}
+}
+
+func TestExpandVariable(t *testing.T) {
+	context := map[string]interface{}{
+		"bool":   true,
+		"num":    float64(99),
+		"string": "abc",
+		"array":  []interface{}{float64(1), float64(2), float64(3), "a", "b", "c"},
+		"obj": map[string]interface{}{
+			"a": float64(1),
+			"b": "abcd",
+		},
+		"null": nil,
+	}
+
+	tests := []struct {
+		variable   string
+		expected   interface{}
+		expectList bool
+	}{
+		{"$('')", "", false},
+		{"$(bool)", context["bool"], false},
+		{"$(['bool'])", context["bool"], false},
+		{"$(num)", context["num"], false},
+		{"$(string)", context["string"], false},
+		{"$(array)", context["array"], false},
+		{"$(obj)", context["obj"], false},
+		{"$(obj.a)", context["obj"].(map[string]interface{})["a"], false},
+		{"$(['obj'].a)", context["obj"].(map[string]interface{})["a"], false},
+		{"$(['obj']['a'])", context["obj"].(map[string]interface{})["a"], false},
+		{"$(obj['a'])", context["obj"].(map[string]interface{})["a"], false},
+		{"$(null)", context["null"], false},
+		{"$(array[*])", context["array"], true},
+		{"$(array[1:3])", []interface{}{context["array"].([]interface{})[1], context["array"].([]interface{})[2]}, true},
```

Comment:
Minor nit: suggest a helper func like:

```go
func arrLookup(context map[string]interface{}, i int) interface{} {
  return context["array"].([]interface{})[i]
}
```

Just to make this expected values here and line 151 a tad more readable.

Reply:
sure. will do.